src/ShiftRegister74HC595.cpp

```diff
@@ -14,46 +14,58 @@ ShiftRegister74HC595::ShiftRegister74HC595(int numberOfShiftRegisters, int seria
 {
     // set attributes
     _numberOfShiftRegisters = numberOfShiftRegisters;
-    
+
     _clockPin = clockPin;
     _serialDataPin = serialDataPin;
     _latchPin = latchPin;
-    
+
     // define pins as outputs
     pinMode(clockPin, OUTPUT);
     pinMode(serialDataPin, OUTPUT);
     pinMode(latchPin, OUTPUT);
-    
+
     // set pins low
     digitalWrite(clockPin, LOW);
     digitalWrite(serialDataPin, LOW);
     digitalWrite(latchPin, LOW);
-    
+
     // allocates the specified number of bytes and initializes them to zero
     _digitalValues = (uint8_t *)malloc(numberOfShiftRegisters * sizeof(uint8_t));
     memset(_digitalValues, 0, numberOfShiftRegisters * sizeof(uint8_t));
-    
-    setAll(_digitalValues); // reset shift register
+
+    updateRegisters();       // reset shift register
 }
 
+// ShiftRegister74HC595 destructor
+// The memory allocated in the constructor is also released.
+ShiftRegister74HC595::~ShiftRegister74HC595()
+{
+    free(_digitalValues);
+}
 
 // Set all pins of the shift registers at once.
 // digitalVAlues is a uint8_t array where the length is equal to the number of shift registers.
-void ShiftRegister74HC595::setAll(uint8_t * digitalValues)
+void ShiftRegister74HC595::setAll(const uint8_t * digitalValues)
 {
-    int byte;
-    
-    for (byte = _numberOfShiftRegisters - 1; byte >= 0; byte--) {
-        shiftOut(_serialDataPin, _clockPin, MSBFIRST, digitalValues[byte]);
-    }
-    
-    _digitalValues = digitalValues; 
-    
-    digitalWrite(_latchPin, HIGH); 
-    digitalWrite(_latchPin, LOW); 
+    memcpy( _digitalValues, digitalValues, _numberOfShiftRegisters);   // dest, src, size
+    updateRegisters();
 }
 
 
+// Experimental
+// The same as setAll, but the data is located in PROGMEM
+// For example with:
+//     const uint8_t myFlashData[] PROGMEM = { 0x0F, 0x81 };
+#ifdef __AVR__
+void setAll_P(const uint8_t * digitalValuesProgmem)
+{
+    PGM_VOID_P p = reinterpret_cast<PGM_VOID_P>(digitalValuesProgmem);
+    memcpy_P( _digitalValues, p, _numberOfShiftRegisters);
+    updateRegisters();
+}
+#endif
```

Comment:
I have tried to compile with the Arduino IDE version 1.8.9 on Mac, for Arduino Uno and am getting the following error:

```
src/ShiftRegister74HC595.cpp: In function 'void setAll_P(const uint8_t*)':
src/ShiftRegister74HC595.cpp:63:15: error: '_digitalValues' was not declared in this scope
     memcpy_P( _digitalValues, p, _numberOfShiftRegisters);
               ^
src/ShiftRegister74HC595.cpp:63:34: error: '_numberOfShiftRegisters' was not declared in this scope
     memcpy_P( _digitalValues, p, _numberOfShiftRegisters);
                                  ^
src/ShiftRegister74HC595.cpp:64:21: error: 'updateRegisters' was not declared in this scope
     updateRegisters();
                     ^
src/ShiftRegister74HC595.cpp: In member function 'void ShiftRegister74HC595::updateRegisters()':
src/ShiftRegister74HC595.cpp:90:55: error: 'digitalValues' was not declared in this scope
         shiftOut(_serialDataPin, _clockPin, MSBFIRST, digitalValues[i]);
                                                       ^
exit status 1
Error compiling for board Arduino/Genuino Uno.
```